qiskit_addon_sqd/fermion.py

```diff
@@ -66,6 +79,351 @@ def load(cls, filename):
         with np.load(filename) as data:
             return cls(data["amplitudes"], data["ci_strs_a"], data["ci_strs_b"])
 
+    def rdm(self, rank: int = 1, spin_summed: bool = False) -> np.ndarray:
+        """Compute reduced density matrix."""
+        # Reason for type: ignore: mypy can't tell the return type of the
+        # PySCF functions
+        sci_vector = _as_SCIvector(self.amplitudes, (self.ci_strs_a, self.ci_strs_b))
+        if rank == 1:
+            if spin_summed:
+                return make_rdm1(sci_vector, self.norb, self.nelec)  # type: ignore
+            return make_rdm1s(sci_vector, self.norb, self.nelec)  # type: ignore
+        if rank == 2:
+            if spin_summed:
+                return make_rdm2(sci_vector, self.norb, self.nelec)  # type: ignore
+            return make_rdm2s(sci_vector, self.norb, self.nelec)  # type: ignore
+        raise NotImplementedError(
+            f"Computing the rank {rank} reduced density matrix is currently not supported."
+        )
+
+    def orbital_occupancies(self) -> tuple[np.ndarray, np.ndarray]:
+        """Average orbital occupancies."""
+        dm_a, dm_b = self.rdm(rank=1, spin_summed=False)
+        return np.diagonal(dm_a), np.diagonal(dm_b)
+
+
+@dataclass(frozen=True)
+class SCIResult:
+    """Result of an SCI calculation."""
+
+    energy: float
+    """The SCI energy."""
+
+    sci_state: SCIState
+    """The SCI state."""
+
+    orbital_occupancies: tuple[np.ndarray, np.ndarray]
+    """The average orbital occupancies."""
+
+    rdm1: np.ndarray | None = None
+    """Spin-summed 1-particle reduced density matrix."""
+
+    rdm2: np.ndarray | None = None
+    """Spin-summed 2-particle reduced density matrix."""
+
+
+def diagonalize_fermionic_hamiltonian(
+    one_body_tensor: np.ndarray,
+    two_body_tensor: np.ndarray,
+    bit_array: BitArray,
+    samples_per_batch: int,
+    norb: int,
+    nelec: tuple[int, int],
+    *,
+    num_batches: int = 1,
+    energy_tol: float = 1e-8,
+    occupancies_tol: float = 1e-5,
+    max_iterations: int = 100,
+    sci_solver: Callable[
+        [list[tuple[np.ndarray, np.ndarray]], np.ndarray, np.ndarray, int, tuple[int, int]],
+        list[SCIResult],
+    ]
+    | None = None,
+    symmetrize_spin: bool = False,
+    include_configurations: list[int] | tuple[list[int], list[int]] | None = None,
+    initial_occupancies: tuple[np.ndarray, np.ndarray] | None = None,
+    carryover_threshold: float = 1e-4,
+    callback: Callable[[list[SCIResult]], None] | None = None,
+    seed: int | np.random.Generator | None = None,
+) -> SCIResult:
+    """Run SQD.
+
+    Args:
+        one_body_tensor: The one-body tensor of the Hamiltonian.
+        two_body_tensor: The two-body tensor of the Hamiltonian.
+        bit_array: Array of sampled bitstrings. Each bitstring should have both the
```

Comment:
Could we give a concrete example since bit ordering is always tricky/arbitrary? (e.g. `[q0_b, ..., qN_b, q0_a, ...qN_b]`)

Reply:
Done.